checksum/checksum.go

```diff
@@ -0,0 +1,180 @@
+package checksum
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/schollz/progressbar/v3"
+)
+
+func checksumFor(path string) ([]byte, error) {
+	h := md5.New()
+
+	var r *os.File
+	r, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file in %s: %w", path, err)
+	}
+	defer r.Close()
+
+	if _, err := io.Copy(h, r); err != nil {
+		return nil, fmt.Errorf("error reading %s checksum content: %w", path, err)
+	}
+
+	return h.Sum(nil), nil
+}
+
+func check(src, target, f string) (bool, error) {
+	name := filepath.Base(f)
+	sc, err := checksumFor(filepath.Join(src, name))
+	if err != nil {
+		return false, fmt.Errorf("error getting src %s checksum: %w", name, err)
+	}
+
+	tc, err := checksumFor(filepath.Join(target, name))
+	if err != nil {
+		return false, fmt.Errorf("error getting target %s checksum: %w", name, err)
+	}
+
+	return bytes.Equal(sc, tc), nil
+}
+
+// Check compares md5 checksum files from src directory with corresponding files in target directory
+func Check(src, target string) error {
+	ls, err := filepath.Glob(filepath.Join(src, "*.md5"))
+	if err != nil {
+		return fmt.Errorf("error listing checksum files in %s directory: %w", src, err)
+	}
+
+	if len(ls) == 0 {
+		return fmt.Errorf("target directory %s has no checksum files to compare with", target)
+	}
+
+	bar := progressbar.Default(int64(len(ls)))
+	defer bar.Close()
+	bar.Describe("Checking files checksum")
+	if err := bar.RenderBlank(); err != nil {
+		return fmt.Errorf("error rendering the progress bar: %w", err)
+	}
+
+	type qm struct {
+		f     string
+		equal bool
+		err   error
+	}
+	q := make(chan qm, len(ls))
+	defer close(q)
+
+	var wg sync.WaitGroup
+
+	for _, f := range ls {
+		wg.Add(1)
+
+		go func(f string, wg *sync.WaitGroup, q chan<- qm) {
+			defer wg.Done()
+
+			equal, err := check(src, target, f)
+
+			q <- qm{f, equal, err}
+		}(f, &wg, q)
+	}
+
+	wg.Wait()
+
+	var notEqual []string
+	for m := range q {
```

Comment:
Se utilizarmos isso, nãop precisamos do `WaitGroup`:

```suggestion
	for i := 0; i < len(ls); i++ {
```

Reply:
Poderia explicar mais? Sem o `WaitGroup` ocorrou o mesmo erro que acontecia com o `sample`, no primeiro erro não nulo que retorna a execução para e as outras go rotinas tentam enviar em canais fechados. A única forma que eu vejo de não usarmos o `WaitGroup` é tratar cada erro específicamente sabendo onde tem que interromper a execução (e arrumar uma forma de interromper a execução de outras go rotinas), e onde tem que continuar.

Mas acho que entre ter esse trabalho, e manter o `WaitGroup`, melhor manter. Você tem alguma outra idéia? Não consegui entender muito bem como esse `for ...` substituiria.

Reply:
Pode manter, isso é mais um detalhe mesmo : )